include/roboteam_utils/LineSegment.h

```diff
@@ -1,249 +1,132 @@
+         * @return The projection point. Note that this point is located on this LineSegment, but it does NOT have to be orthogonal to the given point.
+         */
+        [[nodiscard]] Vector2 project(const Vector2 &point) const;
+
+        /**
+         * Get the intersection between two (finite) LineSegment. Note that this function knows how to deal with LineSegments that are points and how to deal with parallel
+         * LineSegments. Moreover this function guarantees to return an intersection point if the LineSegments intersects. In case:
+         * - There is no intersection then std:nullopt is returned.
+         * - There is a single intersection point then that intersection point is returned.
+         * - There are multiple intersection points (infinitely many) which happens when the LineSegments overlap then it prefers to returns 1: start, 2: end, 3: line.start of which
+         *  the first most option is selected that lies on both LineSegments. Note that if none of these points lie on both LineSegments and if both the LineSegments are parallel then
+         *  the LineSegments cannot intersect.
+         *
+         * @param line The other (finite) LineSegment.
+         * @return std::nullopt if the LineSegments do not intersect. Otherwise return an intersection point.
+         */
+        [[nodiscard]] std::optional<Vector2> intersects(const LineSegment &line) const;
+
+        /**
+         * Check if two LineSegment intersect. Note that this function knows how to deal with LineSegments that are points and how to deal with parallel LineSegments.
+         *
+         * @param line The other (finite) LineSegment.
+         * @return True if the LineSegments have any type of intersection (overlap is also considered as intersection) and false otherwise.
+         */
+        [[nodiscard]] bool doesIntersect(const LineSegment &line) const;
+
+        /**
+         * Return the intersection(s) of two LineSegments. This is function is similar to the intersects function in the sense that it knows how to deal with parallel LineSegments and
+         * LineSegments that are actually points. But it is different from the intersects function in the sense that it can return up to 2 intersections. In case:
+         * - The LineSegments do not intersect, it returns an empty list.
+         * - The LineSegments do only intersect once, it returns a list with only that intersection point.
+         * - The LineSegments intersect multiple times (infinitely often), then these intersections form a LineSegment L. In this case it returns the start and end points of
+         *  LineSegment L.
+         *
+         * @param line The other (finite) LineSegment.
+         * @return Returns a list of the intersections.
+         */
+        [[nodiscard]] std::vector<Vector2> multiIntersect(const LineSegment &line) const;
+
+        /**
+         * Destructor of the LineSegment.
+         */
+        virtual ~LineSegment() = default;
```

Comment:
Why is there a virtual destructor here? Am I missing something?

Reply:
I did not add or remove the destructor. I only changed the ordening of the functions.